Marlin/src/HAL/STM32/can/FDCAN_toolhead.cpp

```diff
@@ -0,0 +1,733 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (c) 2025 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (c) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+/**
+ * Contributor Notes:
+ * ===============
+ * In \Users\<username>\.platformio\packages\framework-arduinoststm32@4.20600.231001\libraries\SrcWrapper\src\HardwareTimer.cpp
+ * Add "__weak" in front of "void TIM16_IRQHandler(void)" to be able to override the function and redirect the interrupt
+ */
+
+#include "../../../inc/MarlinConfigPre.h"
+
+#if ENABLED(CAN_TOOLHEAD)
+// TODO:
+// 1. Convert into class
+// 2. Pickup correct CAN device from CAN_RX_PIN/CAN_TX_PIN
+// 3. Calculate baudrate and data sampling timing from clock frequencies defined by CAN_BAUDRATE
+
+#include "../../platforms.h"
+#include "../../../gcode/gcode.h"
+#include "../../../gcode/parser.h"
+#include "../../../gcode/queue.h"
+#include "../../../module/temperature.h"
+#include "../../../libs/numtostr.h"
+#include "../../../inc/MarlinConfig.h"
+#include "../../../feature/controllerfan.h"
+#include "../../../core/serial.h"
+
+#include "../../shared/CAN.h"
+
+#define DEBUG_OUT ENABLED(CAN_DEBUG)
+#include "../../../core/debug_out.h"
+
+extern "C" void TIM16_IRQHandler(void); // Override weak functions
+extern "C" void HAL_FDCAN_RxFifo0Callback(FDCAN_HandleTypeDef *hfdcan, uint32_t RxFifo0ITs);
```

Comment:
There are no such types as `FDCAN_HandleTypeDef` or `FDCAN_RxHeaderTypeDef`, and no such function as `HAL_FDCAN_GetRxMessage`. But if I alter the types to `CAN_HandleTypeDef` and `CAN_RxHeaderTypeDef` then the build fails because there is no such field as `CAN_rx_header.IdType`.